cardinal/router/router_evm_server.go

```diff
@@ -30,13 +33,44 @@ type evmServer struct {
 
 	provider   Provider
 	grpcServer *grpc.Server
+	routerKey  string
 }
 
-func newEvmServer(p Provider) *evmServer {
-	return &evmServer{
-		provider:   p,
-		grpcServer: grpc.NewServer(),
+func newEvmServer(p Provider, routerKey string) *evmServer {
+	e := &evmServer{
+		provider:  p,
+		routerKey: routerKey,
 	}
+	e.grpcServer = grpc.NewServer(grpc.UnaryInterceptor(e.serverCallInterceptor))
+	return e
+}
+
+// serverCallInterceptor catches calls to handlers and ensures they have the right secret key.
+func (e *evmServer) serverCallInterceptor(
+	ctx context.Context,
+	req any,
+	_ *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (resp any, err error) {
+	// we only want to guard the SendMessage method. not the query shard method.
+	if _, ok := req.(*routerv1.SendMessageRequest); !ok {
+		return handler(ctx, req)
+	}
+	md, ok := metadata.FromIncomingContext(ctx)
```

Comment:
You can use the router key package/type mentioned above to add `jwt.FromIncomingContext` to obtain the token from the request context without having to hardcode "router-key" here.

Reply:
I don't think this is resolved yet. I don't think this is resolved yet.

You can just add `FromIncomingContext` to the credentials package so you don't have to do `routerKey[md.credentials.TokenKey]`

Reply:
just to clarify, you want the credentials package to extract the key from the ctx and return it to the `serverCallInterceptor` method?